tests/generic-java-signatures/outer-ref-elimination.scala

```diff
@@ -0,0 +1,36 @@
+// This test checks that references to outer classes in inner classes are
+// eliminated in some cases when they are not used. This is done is the
+// ExplicitOuter phase. See issue #19569 for discussions.
+
+object helper {
+  def test(cls: Class[?]) = println(cls.getDeclaredConstructors.toList)
+}
+
+import helper.test
+
+object T1 { class C1; test(classOf[C1]) }
+object T2 { new AnyRef { class C2; test(classOf[C2]) } }
+object T3 { def t3(): Unit = { class C3; test(classOf[C3]) } }
+object T4 { def t4(): Unit = new AnyRef { class C4; test(classOf[C4]) } }
+
+class T5 { class C5; test(classOf[C5]) } // outer ref currently not elided
```

Comment:
I think "currently" isn't correct (and thus misleading).

I think the term "referenced" might be overused.

```scala
  /** Class needs an outer pointer, provided there is a reference to an outer this in it. */
  def needsOuterIfReferenced(cls: ClassSymbol)(using Context): Boolean =
```

Here it's about internal reference to the outer enclosing class.

```scala
    (!hasLocalInstantiation(cls) || // needs outer because we might not know whether outer is referenced or not
```

Here I think "is referenced" means "we can't tell that the user has no stable reference to the outer".

Here's the counter-example of why we should _not_ make C5 lose its outer pointer:

```scala
    val a = new T5
    val b = new T5
    val x = new a.C5
    val y = new b.C5
    println(x == y)
```

Even though C5 has no reference in it to its enclosing T5, I think it would be incorrect for x and y to be equal.